content/docs/user-guide/data-pipelines/defining-pipelines.md

```diff
@@ -0,0 +1,237 @@
+
+</admon>
+
+Each stage wraps around an executable shell [command] and specifies the
+necessary inputs as well as expected [outputs](#outputs) (if any). Let's look at
+a sample stage that [depends](#simple-dependencies) on a script file it runs as
+well as on a raw data directory it processes (ideally [tracked by DVC] already):
+
+```yaml
+stages:
+  prepare:
+    cmd: source src/cleanup.sh
+    deps:
+      - src/cleanup.sh
+      - data/raw
+    outs:
+      - data/clean.csv
+```
+
+<admon type="info">
+
+We use [GNU/Linux](https://www.gnu.org/software/software.html) in these
+examples, but Windows or other shells can be used too.
+
+</admon>
+
+Besides writing `dvc.yaml` files manually (recommended), you can also create
+stages with `dvc stage add` -- a limited command-line interface to setup
+pipelines. Let's add another stage this way and look at the resulting
+`dvc.yaml`:
+
+```dvc
+$ dvc stage add --name train \
+                --deps src/model.py \
+                --deps data/clean.csv \
+                --outs data/predict.dat \
+                python src/model.py data/clean.csv
+```
+
+```yaml
+stages:
+  prepare:
+    ...
+    outs:
+      - data/clean.csv
+  train:
+    cmd: python src/model.py data/model.csv
+    deps:
+      - src/model.py
+      - data/clean.csv
+    outs:
+      - data/predict.dat
+```
+
+<admon type="tip">
+
+One advantage of using `dvc stage add` is that it will verify the validity of
+the arguments provided (otherwise stage definition won't be checked until
+execution). A disadvantage is that some advanced features such as [templating]
+are not available this way.
+
+[command]: /doc/user-guide/project-structure/dvcyaml-files#stage-commands
+[templating]: /doc/user-guide/project-structure/pipelines-files#templating
+
+</admon>
+
+Notice that the new `train` stage depends on the output from stage `prepare`
+(`data/clean.csv`), forming the [DAG](#directed-acyclic-graph-dag).
+
+[tracked by dvc]: /doc/start/data-management
+
+## Simple dependencies
+
+There's more than one type of stage dependency. A simple dependency is a file or
+directory used as input by the stage command. When it's contents have changed,
+DVC "invalidates" the stage -- it knows that it needs to run again. This in turn
+may cause a chain reaction in which subsequent stages of the
+<abbr>pipeline</abbr> are also reproduced.
+
+<admon type="info">
+
+DVC [calculates a hash] of file/dir contents to compare vs. previous versions.
+This is a distinctive mechanism over traditional build tools like `make`.
+
+[calculates a hash]:
+  /doc/user-guide/project-structure/internal-files#structure-of-the-cache-directory
+
+</admon>
+
+File system-level dependencies are defined in the `deps` list of `dvc.yaml`
+stages; Alternatively, using the `--deps` (`-d`) option of `dvc stage add` (see
+the previous section's example).
+
+<details>
+
+### External dependencies: click to learn more.
```

Comment:
not sure we call imports external dependencies tbh

Reply:
They're literally `deps` to en external path or URL in `dvc.yaml`.

Reply:
kk, yes, they are probably all ext deps indeed